Qt: Add network port input box to GUI settings

hsjoberg opened the pull request:
Related issue #7039

hsjoberg committed:
Qt/Options: Adding network port to GUI settings, fixes #7039

Adds Network port input box to the Network tab in the Options dialog.
-port takes priority over the GUI setting.
If left blank, it will default to the default port

tulip0 commented:
It's counterintuitive to most people that Bitcoin Core will refuse to use non-standard ports as anything but a last resort, does the option label need a GUI label or a confirmation popup to reflect this? I suspect a lot of people will be mystified otherwise that they don't get any incoming connections after changing a fairly innocuous looking option.

jonasschnelli commented on src/qt/optionsdialog.cpp:
```diff
@@ -45,6 +45,8 @@ OptionsDialog::OptionsDialog(QWidget *parent, bool enableWallet) :
     ui->threadsScriptVerif->setMinimum(-GetNumCores());
     ui->threadsScriptVerif->setMaximum(MAX_SCRIPTCHECK_THREADS);
 
+    ui->networkPort->setValidator(new QIntValidator(1, 65535, this));
```
We need a better protection to not run into a startup termination if the user did enter a invalid port. IIRC the `QIntValidator` still allows one to enter `0` as port number which will very likely result in a locked Bitcoin-QT (startup will stop because of a invalid port). Haven't tested although.
Maybe check: https://github.com/bitcoin/bitcoin/pull/7025

tulip0 replied:
This should probably set 1-1023 as invalid as well as only root can bind them.

luke-jr replied:
@tulip0 That is not universally true. I think Windows allows non-admin to bind them, and on Linux there are capabilities to allow binding them.

jonasschnelli commented:
Concept ACK.

We just need to make sure, people can only enter reasonable/valid ports. Because if someone enters port 80, bitcoin-qt is very likely "bricked" (unless he start with root permissions!) until he manages to startup Bitcoin-Qt with a override port setting (through a shell with `-port=8333' or by changing bitcoin.conf).

Best would be to ask the user – during startup – if he likes to switch to the standard port if the given port can't be used (permissions or already used through another process).

jonasschnelli commented:
https://bitcoin.jonasschnelli.ch/pulls/7107/

Cocosoft commented:
Thanks for the feedback @jonasschnelli, I'll work on the things you pointed out.  
I'll be available more and more in #bitcoin #bitcoin-dev etc as cocoBTC.

fanquake commented:
@Cocosoft Are you still working on this?

Cocosoft commented:
@fanquake Yes, I do intend to continue working on this.

tulip0 commented:
Again, this needs a warning that users will get essentially no incoming connections if they change the port. It's not initiative behaviour and will cause significant confusion, I'm not convinced that this option should even be exposed to users. The address rumouring system really can't be used in conjunction with anything but a single static port.

laanwj commented:
Concept ACK.

Agree that adding a warning (e.g. as mouseover) would make sense, and that ports should be restricted to valid ports (>1024 at least).

Cocosoft commented:
Just FYI, I still intend to commit to this issue. I am still alive.

@laanwj, agreed.

luke-jr committed:
Qt/Options: Avoid changing names of unrelated widgets

luke-jr committed:
Qt/Options: Don't allow setting a port below 1024

luke-jr committed:
Qt/Options: Actually check validator acceptability for network port

luke-jr committed:
Qt/Options: Allow an "invalid" network port if it is the current value already

luke-jr committed:
Qt/NetworkStyle: Use different QSettings for regtest vs testnet

luke-jr commented:
@Cocosoft I've addressed the port number concern, and modified NetworkStyle to store regtest settings separately from testnet.

```
git pull https://github.com/luke-jr/bitcoin.git qtnetworkports
```

Cocosoft commented:
@luke-jr, great, thank you for your help!

Cocosoft commented on src/qt/optionsmodel.cpp:
```diff
@@ -246,6 +254,18 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
+            if (settings.value("nNetworkPort") != value) {
+                // If the port input box is empty, set to default port
+                if (value.toString().isEmpty()) {
+                    settings.setValue("nNetworkPort", (quint16)Params().GetDefaultPort());
```
Maybe I'm mistaken, but won't this code never be reached because of acc8f4e?

luke-jr replied:
You mean because the empty field is invalid input? I suppose so.

jonasschnelli commented:
Needs rebase.
I could continue this. I think it needs some additional "focus-lost" checks (also see how proxy port settings do work) and I could imaging a problem when setting the port to a port-number that is already in use (could prevent from starting Bitcoin-Qt).

jonasschnelli commented:
@ping Cocosoft. Are you interested to finish this?

Cocosoft commented:
@jonasschnelli, I am interested in finishing this. I do understand however if you or someone else wants to take over because it has been idling for so long.

I'll try to be on the IRC channel this week.

laanwj commented:
> I could imaging a problem when setting the port to a port-number that is already in use (could prevent from starting Bitcoin-Qt).

Possibly in the case of the GUI it could be non-fatal if binding a port at startup fails. After warning the user there's no harm in continuing with the port closed.

This would make this change a lot less dangerous.

arowser commented:
Can one of the admins verify this patch?

fanquake commented:
@Cocosoft Are you still planning on finishing this? This needs a rebase at least.

Cocosoft commented:
@fanquake Yes, I haven't forgotten about this and I still plan to continue, I just need to find time. I fully understand if someone wants to finish it up though, as it has been delayed for so long.

luke-jr commented:
Why did you add an enableOkButton and disableOkButton methods that are unused?

Cocosoft commented:
I did not, actually you made them according to Git:

```bash
git blame -L 227,227 src/qt/optionsdialog.cpp
578064a6 (Luke Dashjr 2016-02-12 20:08:00 +0000 227) void OptionsDialog::enableOkButton()
```

I'll rebase again and fix that.

luke-jr commented:
That code wasn't part of my commit; somehow it got added in when you rebased. Perhaps an accident during a merge?

Cocosoft commented:
You're right, sorry. It seems like Phillip Kaufman wrote the functions back in 2013:

```bash
$ git blame -L 227,227 src/qt/optionsdialog.cpp
7e195e84 (Philip Kaufmann 2013-12-03 09:10:10 +0100 227)     setOkButtonState(false);
```

Cocosoft commented:
I'm not sure where the functions come from, I can confirm (by a backup) that they existed before the rebase I did.

I'll remove them.

fanquake commented:
This needs a rebase.

hsjoberg committed:
Adding functions SetArg and SetBoolArg to util.cpp

Lets you override an argument even if it has already been set,
in contrast to SoftSetArg/SoftSetBoolArg.

hsjoberg committed:
Qt: Ask user to use standard port on startup if specified port is in use

If the port (specified by -port or GUI setting) is already in use when
starting the GUI client, and it's not the standard port, ask the user if
they want to listen via the standard network port instead.

Cocosoft commented:
As per @jonasschnelli suggestion, I added a Question box on startup to ask if the user wants to use the standard network port if the specified port (whether it's from -port or in GUI) is already in use.

jonasschnelli commented:
Needs rebase and nit fixes.

jonasschnelli commented:
Closing due to inactivity. Happy to reopen once this has made progress.

Cocosoft commented:
Hi @jonasschnelli, I've rebased the branch upon master on cocosoft/qtnetworkport.
I'm not sure what nits are still missing, but I would be happy to fix them.

I added a warning prompt on startup if the user tries to use a non-standard port that is already in use, and gives the option to use the standard port.

Best
Hampus
